server/src/main/java/io/crate/planner/statement/DeletePlanner.java

```diff
@@ -133,7 +133,8 @@ public void executeOrFail(DependencyCarrier executor,
                 subQueryResults,
                 plannerContext.transactionContext(),
                 executor.nodeContext());
-            if (!where.partitions().isEmpty() && !where.hasQuery()) {
+            if (!where.partitions().isEmpty()
+                && (!where.hasQuery() || WhereClause.canMatch(where.query()))) {
```

Comment:
```suggestion
                && (!where.hasQuery() || Literal.BOOLEAN_TRUE.equals(where.query()))) {
```

I think `canMatch` could also return `true` for queries that have a chance of matching a sub-set of the records. In that case we cannot delete the full partition.


See https://github.com/crate/crate/blob/ced78125bab6bead35129f2d6f1618188a8536e5/server/src/main/java/io/crate/analyze/where/WhereClauseAnalyzer.java#L142-L149 


`test_where_on_date_with_null_partition_or_id_can_match_all_partitions` and `test_where_clause_that_could_match_on_null_partition_filters_correct_records` are examples where that can be the case.

Reply:
Yep, true. WIll check the quality against the Literal.BOOLEAN_TRUE 👍 tnx